README.md
[![Pub](https://img.shields.io/pub/v/input_slider.svg)](https://pub.dev/packages/input_slider)

# input_slider

While Sliders offer a quick and visual way to change a value, it can be hard to achieve precise inputs with them. Additionally, some users really dislike Sliders, and especially on small devices they can get quite fiddly. One way to use the advantages of having a Slider while minimizing the drawbacks is to add an additional alternativ (text) input field.

This packages InputSlider combines a TextField with a Slider and synchronizes both of them. They share the same minimum and maximum values and changes made to one also changes the other. Optionally, you can also specify a leading widget, such as a label Text or Icon. 

![grafik](https://user-images.githubusercontent.com/44811445/118550328-d7c7c580-b75c-11eb-8212-a213eb8afd00.png)


## Usage
```dart 
InputSlider(
  onChange: (value) => print("change: $value"),
  min: 0.0,
  max: 100.0,
  decimalPlaces: 0,
  value: 50.0,
  leading: Text("Percentage:"),
)
```

## Input Slider Forms

You can group multiple InputSlider together using an InputSliderForm. By specifying `leadingWeight` and `sliderWeight` it is possible to align them together. They can be also styled together or individually. 

```dart
InputSliderForm(
    leadingWeight: 1,
    sliderWeight: 20,
    activeSliderColor: Colors.red,
    inactiveSliderColor: Colors.green[100],
    filled: true,
    vertical: true,
    children: [
      InputSlider(
        onChange: (value) {
          print("Setting 1 changed");
        },
        min: 0.0,
        max: 10.0,
        decimalPlaces: 0,
        defaultValue: 5.0,
        leading: Text("Setting 1:"),
      ),
      InputSlider(
        onChange: (value) {
          print("Setting 2 changed");
        },
        min: 0.0,
        max: 1.0,
        decimalPlaces: 3,
        defaultValue: 0.32,
      ),
)
```

Sliders can be rotated by 90 degrees by setting vertical=true. This can be done to individual InputSliders or to InputSliderForms.

![photo_2021-07-10_13-54-59](https://user-images.githubusercontent.com/44811445/125162105-6f0e2f80-e186-11eb-82f6-2fcccbc3e916.jpg)



CHANGELOG.md
## [0.3.1] - 08.10.2022

- Changes textfield keyboard type to allow for decimal point and negative values on iOS.

## [0.3.0] - 10.07.2021

- Adds vertical slider support: Set vertical=true to rotate the Slider by 90 degrees. 

## [0.2.3] - 06.06.2021

- TextEditingController is now properly disposed

## [0.2.2] - 06.06.2021

- Removes unnecessary print statement

## [0.2.1] - 02.06.2021

- Fixes wrong assertion for default value

## [0.2.0] - 01.06.2021

- Adds leadingWeight and sliderWeight 
- Adds InputSliderForm which allows to group, align and style InputSliders together
- BREAKING: Changes value to defaultValue

## [0.1.4] - 17.05.2021

- Fixes bug with onChange Function
- Fixes bug with onChangeEnd Function

## [0.1.3] - 17.05.2021

- Updates description

## [0.1.2] - 17.05.2021

- Adds example

## [0.1.1] - 17.05.2021

- Updates readme
- fillColor works now

## [0.1.0] - 17.05.2021

- Initial release of input_slider


lib/input_slider.dart
library input_slider;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:input_slider/util.dart';

/// An input widget that combines a [Slider] synchronized with a [TextField]
class InputSlider extends StatefulWidget {
  /// Called whenever the value changes by moving the slider or entering a value
  /// into the TextField
  final Function(double) onChange;

  /// Called whenever the user is done moving the slider.
  final Function(double)? onChangeEnd;

  /// Called whenever the user starts moving the slider.
  final Function(double)? onChangeStart;

  /// The minimum value. Every value smaller than this will be clamped.
  final double min;

  /// The maximum value. Every value greater than this will be clamped.
  final double max;

  /// The default value of this InputSlider.
  final double defaultValue;

  /// The number of discrete divisions.
  final int? division;

  /// The color of the active (left) part of the slider.
  ///
  /// By default chooses a color based on the Theme.
  final Color? activeSliderColor;

  /// The color of the inactive (right) part of the slider.
  ///
  /// By default chooses a color based on the Theme.
  final Color? inactiveSliderColor;

  /// The amount of decimal places shown in the TextField.
  final int decimalPlaces;

  /// A leading Widget. This could be a Text or an Icon as a label.
  final Widget? leading;

  /// The [TextStyle] used in the TextField.
  final TextStyle? textFieldStyle;

  /// Whether the TextField is filled.
  /// Ignored if [inputDecoration] is non-null.
  final bool? filled;

  /// The color with which the TextField is filled, if [filled] is true.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? fillColor;

  /// The border color of the TextField if not focused.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? borderColor;

  /// The border color of the TextField if focused.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? focusBorderColor;

  /// The border radius of the TextField.
  /// Ignored if [inputDecoration] is non-null.
  final BorderRadius? borderRadius;

  /// The [InputDecoration] used by the TextField. If null, use a default decoration.
  final InputDecoration? inputDecoration;

  /// Determines the proportional weight (flex) of the leading widget.
  ///
  /// If null or 0, the leading widget's preferred size is used.
  final int? leadingWeight;

  /// Determines the proportional weight (flex) of the slider.
  ///
  /// If 0 the slider's preferred size is used.
  /// If null the slider is extended.
  final int? sliderWeight;

  /// The size of the input [TextField].
  ///
  /// If null, calculates the size based on the [textFieldStyle] and [decimalPlaces].
  final Size? textFieldSize;

  /// If true, rotates the [Slider] by 90 degrees. Keeps the orientation of the
  /// [TextField] and the leading widget. Default is false.
  ///
  /// Note: This puts all widgets ([Slider], [TextField] and leading widget) inside a
  /// Column instead of a row. When using this [InputSlider] inside another Column,
  /// you have to constrain the height of this [InputSlider] by using widgets such
  /// as [SizedBox], [Expanded] or [Flexible].
  final bool vertical;

  const InputSlider(
      {required this.onChange,
      required this.min,
      required this.max,
      required this.defaultValue,
      this.onChangeEnd,
      this.onChangeStart,
      this.leading,
      this.decimalPlaces = 2,
      this.division,
      this.activeSliderColor,
      this.inactiveSliderColor,
      this.textFieldStyle,
      this.filled,
      this.fillColor,
      this.borderColor,
      this.focusBorderColor,
      this.borderRadius,
      this.inputDecoration,
      this.leadingWeight,
      this.sliderWeight,
      this.textFieldSize,
      this.vertical = false})
      : super();

  @override
  _InputSliderState createState() => _InputSliderState(
      defaultValue: defaultValue, textFieldSize: this.textFieldSize);
}

class _InputSliderState extends State<InputSlider> {
  double defaultValue;
  TextEditingController _controller = TextEditingController();
  Size? textFieldSize;

  _InputSliderState({required this.defaultValue, this.textFieldSize});

  @override
  void initState() {
    super.initState();
    assert(defaultValue >= widget.min && defaultValue <= widget.max,
        "value must be between min and max.");
    _controller = TextEditingController(
        text: defaultValue.toStringAsFixed(widget.decimalPlaces));
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (textFieldSize == null) _calculateTextFieldSize();

    final widgets = [
      Flexible(
          flex: widget.leadingWeight ?? 0,
          fit: FlexFit.tight,
          child: Align(
              alignment: Alignment.centerLeft,
              child: widget.leading)),
      Padding(
        padding: EdgeInsets.only(left: 8.0),
      ),
      Padding(
        padding: const EdgeInsets.all(8.0),
        child: SizedBox(
          width: textFieldSize!.width,
          height: textFieldSize!.height,
          child: TextField(
            controller: _controller,
            keyboardType: TextInputType.numberWithOptions(signed: true, decimal: true),
            style: widget.textFieldStyle ?? DefaultTextStyle.of(context).style,
            onSubmitted: (value) {
              double parsedValue = double.tryParse(value) ?? this.defaultValue;
              parsedValue = parsedValue.clamp(widget.min, widget.max);
              setState(() {
                this.defaultValue = parsedValue;
              });
              _setControllerValue(this.defaultValue);
              widget.onChangeEnd?.call(this.defaultValue);
            },
            textAlign: TextAlign.center,
            decoration: widget.inputDecoration ??
                InputDecoration(
                  enabledBorder: OutlineInputBorder(
                      borderRadius:
                          widget.borderRadius ?? BorderRadius.circular(8),
                      borderSide: BorderSide(
                          color: widget.borderColor ??
                              Theme.of(context).hintColor)),
                  border: OutlineInputBorder(
                      borderRadius:
                          widget.borderRadius ?? BorderRadius.circular(8),
                      borderSide: BorderSide(
                          color: widget.borderColor ??
                              Theme.of(context).hintColor)),
                  focusedBorder: OutlineInputBorder(
                      borderRadius:
                          widget.borderRadius ?? BorderRadius.circular(8),
                      borderSide: BorderSide(
                          width: 2,
                          color: widget.focusBorderColor ??
                              Theme.of(context).primaryColor)),
                  filled: widget.filled,
                  fillColor: widget.fillColor,
                  contentPadding: EdgeInsets.only(top: 5),
                ),
          ),
        ),
      ),
      widget.vertical ? Flexible(
          flex: widget.sliderWeight ?? 1,
          fit: FlexFit.tight,
          child: RotatedBox(
            quarterTurns: 1,
            child: Slider(
            value: defaultValue,
            min: widget.min,
            max: widget.max,
            divisions: widget.division,
            activeColor: widget.activeSliderColor,
            inactiveColor: widget.inactiveSliderColor,
            onChangeEnd: widget.onChangeEnd,
            onChanged: (double value) {
              setState(() {
                this.defaultValue = value;
              });
              _setControllerValue(value);
            },
          ),
        ),
      ) : Flexible(
        flex: widget.sliderWeight ?? 1,
        fit: FlexFit.tight,
        child: Slider(
          value: defaultValue,
          min: widget.min,
          max: widget.max,
          divisions: widget.division,
          activeColor: widget.activeSliderColor,
          inactiveColor: widget.inactiveSliderColor,
          onChangeEnd: widget.onChangeEnd,
          onChanged: (double value) {
            setState(() {
              this.defaultValue = value;
            });
            _setControllerValue(value);
          },
        ),
      ),
    ];

    return (widget.vertical)
        ? Column(
            mainAxisSize: MainAxisSize.min,
            children: widgets,
          )
        : Row(
            children: widgets,
          );
  }

  /// Calculates the size of the TextField input.
  void _calculateTextFieldSize() {
    TextStyle style =
        widget.textFieldStyle ?? DefaultTextStyle.of(context).style;
    String measureString = widget.max.toString() + ".";
    for (int i = 0; i < widget.decimalPlaces; i++) measureString += "9";
    final textSize = calculateTextSize(measureString, style);
    textFieldSize = textSize + Offset(8, 4);
  }

  /// Sets the [_controller] to the String representation of [value], clamped and
  /// with max [widget.decimalPlaces].
  void _setControllerValue(double value) {
    value = value.clamp(widget.min, widget.max);
    setState(() {
      _controller.text = value.toStringAsFixed(widget.decimalPlaces);
    });
    widget.onChange.call(value);
  }
}


lib/input_slider_form.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'input_slider.dart';
import 'util.dart';

/// An [InputSliderForm] groups multiple [InputSlider], aligns them nicely and
/// can be used to provide the same styling to all of them.
///
/// If the same styling parameter is set for both the [InputSliderForm] and one of
/// its children, the child value is chosen.
class InputSliderForm extends StatelessWidget {
  /// The [InputSlider] contained in this [InputSliderForm]
  final List<InputSlider> children;

  /// The color of the active (left) part of the slider.
  ///
  /// By default chooses a color based on the Theme.
  final Color? activeSliderColor;

  /// The color of the inactive (right) part of the slider.
  ///
  /// By default chooses a color based on the Theme.
  final Color? inactiveSliderColor;

  /// The [TextStyle] used in the TextField.
  final TextStyle? textFieldStyle;

  /// Whether the TextField is filled.
  /// Ignored if [inputDecoration] is non-null.
  final bool? filled;

  /// The color with which the TextField is filled, if [filled] is true.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? fillColor;

  /// The border color of the TextField if not focused.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? borderColor;

  /// The border color of the TextField if focused.
  /// Ignored if [inputDecoration] is non-null.
  ///
  /// By default chooses a color based on the Theme.
  final Color? focusBorderColor;

  /// The border radius of the TextField.
  /// Ignored if [inputDecoration] is non-null.
  final BorderRadius? borderRadius;

  /// The [InputDecoration] used by the TextField. If null, use a default decoration.
  final InputDecoration? inputDecoration;

  /// Determines the proportional weight (flex) of the leading widget.
  ///
  /// If null or 0, the leading widget's preferred size is used.
  final int? leadingWeight;

  /// Determines the proportional weight (flex) of the slider.
  ///
  /// If 0 the slider's preferred size is used.
  /// If null the slider is extended.
  final int? sliderWeight;

  const InputSliderForm(
      {Key? key,
      required this.children,
      this.leadingWeight,
      this.sliderWeight,
      this.inputDecoration,
      this.borderRadius,
      this.focusBorderColor,
      this.borderColor,
      this.filled,
      this.fillColor,
      this.textFieldStyle,
      this.inactiveSliderColor,
      this.activeSliderColor})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    Size maxTextFieldSize = _calculateMaxTextFieldSize(context: context);

    return Column(
        children: children.map((InputSlider slider) {
      return InputSlider(
        onChange: slider.onChange,
        min: slider.min,
        max: slider.max,
        defaultValue: slider.defaultValue,
        onChangeEnd: slider.onChangeEnd,
        onChangeStart: slider.onChangeStart,
        leading: slider.leading,
        decimalPlaces: slider.decimalPlaces,
        division: slider.division,
        activeSliderColor: slider.activeSliderColor ?? activeSliderColor,
        inactiveSliderColor: slider.inactiveSliderColor ?? inactiveSliderColor,
        textFieldStyle: slider.textFieldStyle ?? textFieldStyle,
        filled: slider.filled ?? filled,
        fillColor: slider.fillColor ?? fillColor,
        borderColor: slider.borderColor ?? borderColor,
        focusBorderColor: slider.focusBorderColor ?? focusBorderColor,
        borderRadius: slider.borderRadius ?? borderRadius,
        inputDecoration: slider.inputDecoration ?? inputDecoration,
        leadingWeight: slider.leadingWeight ?? leadingWeight,
        sliderWeight: slider.sliderWeight ?? sliderWeight,
        textFieldSize: maxTextFieldSize,
      );
    }).toList());
  }

  /// Calculates the size of the TextField input of [slider].
  Size _calculateMaxTextFieldSize({required BuildContext context}) {
    Size maxTextFieldSize = Size(0, 0);
    children.forEach((slider) {
      TextStyle style =
          slider.textFieldStyle ?? DefaultTextStyle.of(context).style;
      String measureString = slider.max.toString() + ".";
      for (int i = 0; i < slider.decimalPlaces; i++) measureString += "9";
      final textSize = calculateTextSize(measureString, style);
      final textFieldSize = textSize + Offset(8, 4);
      if (textFieldSize.width > maxTextFieldSize.width)
        maxTextFieldSize = textFieldSize;
    });

    return maxTextFieldSize;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:input_slider/input_slider.dart';
import 'package:input_slider/input_slider_form.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'InputSlider Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage() : super();

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double _volume = 50;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body:
      Padding(
          padding: const EdgeInsets.all(30.0),
          child: Column(
            children: [
              InputSlider(
                  onChange: (value) {
                    setState(() {
                      _volume = value;
                    });
                  },
                  min: 0.0,
                  max: 100.0,
                  decimalPlaces: 0,
                  defaultValue: _volume,
                  activeSliderColor: Colors.green,
                  inactiveSliderColor: Colors.green[100],
                  leading: Icon(Icons.volume_down)),
              _volumeDisplay(),
              Padding(padding: EdgeInsets.symmetric(vertical: 24), child: Divider(),),
              Expanded(
                child: InputSliderForm(
                  vertical: true,
                  leadingWeight: 1,
                  sliderWeight: 20,
                  activeSliderColor: Colors.red,
                  inactiveSliderColor: Colors.green[100],
                  filled: true,
                  children: [
                    InputSlider(
                      onChange: (value) {
                        print("Setting 1 changed");
                      },
                      min: 0.0,
                      max: 10.0,
                      decimalPlaces: 0,
                      defaultValue: 5.0,
                      leading: Text("Setting 1:"),
                    ),
                    InputSlider(
                      onChange: (value) {
                        print("Setting 2 changed");
                      },
                      min: 0.0,
                      max: 1.0,
                      decimalPlaces: 3,
                      defaultValue: 0.32,
                    ),
                    InputSlider(
                      onChange: (value) {
                        print("Setting 3 changed");
                      },
                      min: 0.0,
                      max: 5.0,
                      decimalPlaces: 1,
                      defaultValue: 4.1,
                        leading: Icon(Icons.perm_data_setting_outlined)
                    ),
                  ],
                ),
              ),
            ],
          )),
    );
  }

  Widget _volumeDisplay() {
    if (_volume < 30) return Text("Turn it up!");
    if (_volume < 70) return Text("Nice volume :-)");
    return Text("Ahhhh! My ears!");
  }
}
